dbcon/mysql/ha_mcs_impl.cpp

```diff
@@ -1898,8 +1899,228 @@ uint32_t doUpdateDelete(THD* thd, gp_walk_info& gwi, const std::vector<COND*>& c
     return rc;
 }
 
+inline bool isSupportedToAnalyze(const execplan::CalpontSystemCatalog::ColType& colType)
+{
+    return colType.isUnsignedInteger() || colType.isSignedInteger();
+}
+
 } //anon namespace
 
+int ha_mcs_impl_analyze(THD* thd, TABLE* table)
+{
+    uint32_t sessionID = execplan::CalpontSystemCatalog::idb_tid2sid(thd->thread_id);
+    boost::shared_ptr<execplan::CalpontSystemCatalog> csc =
+        execplan::CalpontSystemCatalog::makeCalpontSystemCatalog(sessionID);
+
+    csc->identity(execplan::CalpontSystemCatalog::FE);
+
+    auto table_name =
+        execplan::make_table(table->s->db.str, table->s->table_name.str, lower_case_table_names);
+
+    // Skip for now.
+    if (table->s->db.length && strcmp(table->s->db.str, "information_schema") == 0)
+        return 0;
+
+    bool columnStore = (table ? isMCSTable(table) : true);
+    // Skip non columnstore tables.
+    if (!columnStore)
+        return 0;
+
+    execplan::CalpontSystemCatalog::RIDList oidlist = csc->columnRIDs(table_name, true);
+    execplan::MCSAnalyzeTableExecutionPlan::ReturnedColumnList returnedColumnList;
+    execplan::MCSAnalyzeTableExecutionPlan::ColumnMap columnMap;
+
+    // Iterate over table oid list and create a `SimpleColumn` for every column with supported type.
+    for (uint32_t i = 0, e = oidlist.size(); i < e; ++i)
+    {
+        execplan::SRCP returnedColumn;
+        const auto objNum = oidlist[i].objnum;
+        auto tableColName = csc->colName(objNum);
+        auto colType = csc->colType(objNum);
+
+        if (!isSupportedToAnalyze(colType))
+            continue;
+
+        execplan::SimpleColumn* simpleColumn = new execplan::SimpleColumn();
+        simpleColumn->columnName(tableColName.column);
+        simpleColumn->tableName(tableColName.table, lower_case_table_names);
+        simpleColumn->schemaName(tableColName.schema, lower_case_table_names);
+        simpleColumn->oid(objNum);
+        simpleColumn->alias(tableColName.column);
+        simpleColumn->resultType(colType);
+        simpleColumn->timeZone(thd->variables.time_zone->get_name()->ptr());
+
+        returnedColumn.reset(simpleColumn);
+        returnedColumnList.push_back(returnedColumn);
+        columnMap.insert(execplan::CalpontSelectExecutionPlan::ColumnMap::value_type(
+            simpleColumn->columnName(), returnedColumn));
+    }
+
+    // Create execution plan and initialize it with `returned columns` and `column map`.
+    execplan::MCSAnalyzeTableExecutionPlan* caep =
```

Comment:
I missed where you delete this ptr. IMHO this pointer should be std::uniqie_ptr.

Reply:
Thanks, yes, I did introduce a memory leak here.